clients/drcachesim/tools/common/decode_cache.cpp

```diff
@@ -0,0 +1,160 @@
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL VMWARE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+#include "decode_cache.h"
+
+namespace dynamorio {
+namespace drmemtrace {
+
+void
+decode_info_base_t::set_decode_info(
+    void *dcontext, const dynamorio::drmemtrace::_memref_instr_t &memref_instr,
+    instr_t *instr)
+{
+    set_decode_info_derived(dcontext, memref_instr, instr);
+    is_valid_ = true;
+}
+
+bool
+decode_info_base_t::is_valid() const
+{
+    return is_valid_;
+}
+
+instr_decode_info_t::~instr_decode_info_t()
+{
+    if (instr_ != nullptr) {
+        instr_destroy(dcontext_, instr_);
+    }
+}
+
+decode_cache_base_t::~decode_cache_base_t()
+{
+    std::lock_guard<std::mutex> guard(module_mapper_mutex_);
+    if (use_module_mapper_) {
+        --module_mapper_use_count_;
+        if (module_mapper_use_count_ == 0) {
+            // We cannot wait for the static module_mapper_ to be destroyed
+            // automatically because we want to do it before DR's global resource
+            // accounting is done at the end.
+            module_mapper_.reset(nullptr);
+            delete[] modfile_bytes_;
+            modfile_bytes_ = nullptr;
+        }
+    }
+}
+
+std::string
+decode_cache_base_t::init_module_mapper(const std::string &module_file_path,
+                                        const std::string &alt_module_dir)
+{
+    std::lock_guard<std::mutex> guard(module_mapper_mutex_);
+    use_module_mapper_ = true;
+    ++module_mapper_use_count_;
+    if (module_mapper_ != nullptr) {
+        // We want only a single module_mapper_t instance to be
+        // initialized that is shared among all instances of
+        // decode_cache_base_t.
+        return "";
+    }
+    return make_module_mapper(module_file_path, alt_module_dir);
+}
+
+std::string
+decode_cache_base_t::make_module_mapper(const std::string &module_file_path,
+                                        const std::string &alt_module_dir)
+{
+    // Legacy trace support where binaries are needed.
+    // We do not support non-module code for such traces.
+    file_t modfile;
+    std::string error = read_module_file(module_file_path, modfile, modfile_bytes_);
+    if (!error.empty()) {
+        return "Failed to read module file: " + error;
+    }
+    dr_close_file(modfile);
+    module_mapper_ =
+        module_mapper_t::create(modfile_bytes_, nullptr, nullptr, nullptr, nullptr,
```

Comment:
We have to override/replace this code internally: currently we do so in an opcode_mix subclass.  This will thus break internal use without corresponding internal changes: make sure you have a valid and tested path forward there.

Reply:
Still trying to address this, so not requesting a re-review just yet.

Reply:
Done.